[ENH] Add LogLoss to TransformedTargetRegressor (TTR)

joshdunnlime opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://skpro.readthedocs.io/en/latest/contribute.html
-->

#### Reference Issues/PRs
#601
Problem Summary: LogLoss not included due to lack of a `log_pdf`. It can be computed but requires calculating the Jacobian of the transform. For many transformers this is non-trivial.

#### What does this implement/fix? Explain your changes.

I have implemented several pieces here. They breakdown as follows:
1) Add `log_pdf` method to TTR.
2) Add `_jacobian` method to TransformedDistribution (TD).
3) Replace `transform` method/function with `transformer` class in the TD.
4) Add `ordered_gradient` function to handle `gradient(f, x)` where the arrays are not sorted by x.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

1) Correctness of numerical calculations in `log_pdf` in TTR.
2) Edge-case of the Jacobian on TD.
3) Does replacing transform with transformer have any unintended consequences or cause a considerable breaking change?

#### Did you add any tests for the change?

I haven't yet added tests. I have modified the `get_test_params` to take a transformer instead of a single function.

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).


<!--
Thanks for contributing!
-->

joshdunnlime committed:
feat: replace inverse with whole transformer on transformed distribution.

This allows us to get more information from the tranformer class. E.g. MinMaxScaler scale_ parameter.

Though not part of the typical forward-facing interface, TransformedDistribution.transform is now no longer callable. Instead TransformedDistribution.transform.inverse_transform would be needed.

joshdunnlime committed:
feat: allows `n_trails=0` for fast optimisation free training

This is useful when testing XGBoostLLS and hyperparameter optimisation is not important. XGBoostLLS can be computationally expensive when searching over the default 30 trials.

joshdunnlime committed:
Merge branch 'xgboostlss-no-trails' into ttr-logloss

joshdunnlime committed:
feat: adds log_pdf to the TargetTranformRregressor

Does so by adding the log_pdf with change-of-variables (jacobian of inverse transform) and direct method for linear transforms, and numerical method for non-linear transforms.

TransformedDistribution now take transformer instead of transform - that is, the transformer class instead of just the inverse transform function/method.

joshdunnlime committed:
docs: Add notebook to show jacobian calcs

This is not actually intended as an example. It is a simple way to share findings. This can (should) be removed before any final merges into main.

joshdunnlime committed:
test: fix TransformdDistribution test with np.exp as inverse_func

joshdunnlime committed:
Revert "Merge branch 'xgboostlss-no-trails' into ttr-logloss"

This reverts commit 4146a6f87a2bd645abc6164fa015159050683406, reversing
changes made to 8bc58970ed09fba334f323971226654b45d0beb1.

fkiraly commented on skpro/distributions/trafo/_transformed.py:
```diff
@@ -62,14 +64,16 @@ class TransformedDistribution(BaseDistribution):
         self,
         distribution,
-        transform,
+        transformer,
```
sorry, we cannot change the names of already released arguments, or it might break user code. Not without deprecation and warnings ate least.

fkiraly commented on skpro/distributions/trafo/_transformed.py:
```diff
@@ -133,7 +199,7 @@ def _ppf(self, p):
             )
 
-        trafo = self.transform
+        trafo = self.transformer.inverse_transform
```
why are we changing this?

joshdunnlime replied:
The `self.transform` is just the inverse function at present, not the whole tranformer class. The idea was to expose the whole class so we can do numerical differentiation or access the `scale_` attr.

fkiraly replied:
I think not easily possible, because the `__init__` resets every estimator that is a component to a pre-fitted state on construction, and because of deprecation.

Overall, I think it is a very good idea, though one may have to find the right implementation.

Would passing `inverse_transform` as an optional additional argument work?

Alternatively, we have to pass something like `sklearn` `FrozenEstimator` wrapping the transformer, though it feels like a bit of an overhead software engineering wise.
The general question is to whether, and if yes how, to allow, API-wise, fitted estimators to be passed to `__init__` - currently this violates the `scikit-learn`-like design.

fkiraly reviewed (changes requested):
Great idea! Though we cannot merge this as is, since:

* this changes the arguments of existing released classes, possibly breaking user code
* I think it introduces too much coupling between TTR and the `TransformedDistribution`, we should avoid that.

Suggestion: how about we allow the user to pass both `transform` and `inverse_transform` (optional) to the `TransformedDistribution`? Then we can do numerical differentiation.

joshdunnlime committed:
docs: fix doctest example for TransformedDistribution

joshdunnlime committed:
feat: allows use of user defined `tranform_diff` function on an skpro transformer

joshdunnlime committed:
feat: add DifferentiableTransformer and base classes

fkiraly reviewed (changes requested):
may I request:

* remove the `BaseTransformer` from this pull request and open a separate one, since we will have some API discussions; let's aim to complete this with numerical differentiation, and later extend it with exact jacobians in a separate PR
* may I suggest to move the numerical differentiation to TD for it to happen only inside `_pdf` and `_log_pdf`? And leave TTR untouched? This we could merge quickly, and it could be "version 0.1" of the feature you want.e
    * testing it might also highlight aspects we have not yet thought about

joshdunnlime committed:
feat: Implement more of the sklearn -> skpro transformer logic

fkiraly commented:
No problem.

May I suggest to change this PR to revert changes across the classes, and add only an approximate `log_pdf` and `pdf` to the TD?

joshdunnlime committed:
add pdf and keep log_pdf

joshdunnlime committed:
remove sample notebook

joshdunnlime committed:
add pdf

fkiraly reviewed (changes requested):
Thanks - almost ready to merge. The public methods should not be overridden, private methods should be implemented.

See how, for instance, `_cdf` is implemented.

joshdunnlime committed:
feat: add approximation of the pdf and log_pdf functions

joshdunnlime committed:
feat: add the _pdf and _log_pdf methods with index and column info

joshdunnlime committed:
fix: change TranformedDistribution measuretype to mixed

joshdunnlime committed:
feat: add index and column information to internal distribution

This is needed so that when TransformerDistribtution.distribution is called, it has the same indices as the wrapper distribution.

joshdunnlime committed:
fix: add pdf and log_pdf to approx to pass pdf_log_pdf test

joshdunnlime commented on skpro/distributions/trafo/_transformed.py:
```diff
@@ -86,6 +88,72 @@ def __init__(
 
         super().__init__(index=index, columns=columns)
 
+    def pdf(self, x: pd.DataFrame):
```
Fixed

fkiraly reviewed (changes requested):
Some questions which may imply changed requests:

* why are we doing the branch "`if hasattr(dist, "_distribution_attr")`? This feels like a violation of the law of demeter. We should only use `dist.pdf`.
* the wrong indices are explained by this, I think. You should not need to correct them if all we do is call `pdf.dist`.

Can you also explain to me how the computation is an approximation to transformed pdf? It is missing the jacobian.

fkiraly commented:
> > I assume this is correct (given the three choices)

> Yes, I think that is the correct choice to hard code, it is the most general class. We can dynamically change this later if we want to.

Actually, if it was `"discrete"`, the code in `__init__` should set it to `"discrete"` via `set_tags`. Otherwise no setting needs to be done, it should be `"mixed"`.

joshdunnlime committed:
use the distribution pdf and log_pdf method (remove scipy workaround)

joshdunnlime commented:
> > > I assume this is correct (given the three choices)
> 
> > Yes, I think that is the correct choice to hard code, it is the most general class. We can dynamically change this later if we want to.
> 
> Actually, if it was `"discrete"`, the code in `__init__` should set it to `"discrete"` via `set_tags`. Otherwise no setting needs to be done, it should be `"mixed"`.

It was `discrete` so I changed it explicitly to `mixed`. It seems to inherit `mixed` from the base class. Should I remove it?

fkiraly commented:
> Would we like to rollback the changes to ttr or are these beneficial? E.g. is a user wants to dig into the internals of TTR/TD the indices are shared between TD and the wrapped distribution.

What are the indices currently? I think they should be the same between row and column index.

fkiraly commented:
I have now opened an additional PR #611, this adds an exact `cdf` for the TD using the transform of the distribution.

Combined with this, #610 should be able to produce a reasonable `pdf` through numerical differentiation.

joshdunnlime commented:
closed in favour of #610 #611 and #612
